pkg/redact/redact.go

```diff
@@ -48,6 +49,37 @@ func Redact(msg interface{}, val reflect.Value) {
 	}
 }
 
+// RedactSensitiveError traverses the given value and processes fields marked as sensitive, redacting their data from error message.
+// Note: The function modifies the error message in place if sensitive data is found.
+func RedactSensitiveError(msg interface{}, val reflect.Value, err *error) {
+	// TODO: This needs to be optimized!
+	// Should cache messages that contain no sensitive data to ignore, and should cache the map of tag number to field name
+
+	if !val.IsValid() {
+		return
+	}
+
+	switch val.Kind() {
+	case reflect.Slice:
+		for i := 0; i < val.Len(); i += 1 {
+			RedactSensitiveError(val.Index(i).Interface(), val.Index(i), err)
+		}
+	case reflect.Map:
+		// TODO: Implement map logic. Currently only certificates in identity use it, and are redacted
+	case reflect.Ptr:
+		RedactSensitiveError(msg, val.Elem(), err)
+	case reflect.Struct:
+		redactErrorMessage(msg, val, err)
+	}
+}
+
+// RedactedError redacts sensitive information from the provided error message.
+// It takes a message of any type and a pointer to an error, and processes the message
+// to remove any sensitive data from error message.
+func RedactedError(msg interface{}, err *error) {
```

Comment:
Thought I added following comment, but couldn't find it.  Adding here again.

This function name is very misleading. It redacts sensitive data from err and doesn't modify msg. Why do we return the cloned message? Why do we even need to clone msg since we are not modifying it? The function name gives the impression that the returned value is an error message with sensitive data redacted, but we are just doing in-place redacting.

Reply:
I removed clone message yesterday